### DISCUSSION TITLE
File handling / Allow to export files

### DISCUSSION THREAD

**Discussion by theelectricsoul** [freeplane/freeplane#1467] [Help]
Hello

Thank you for freeplane :)

# Current functionality around external files( zip, docx, etc.)
* If files are dragged onto the map, an absolute hyperlink is created💗
* If files are dragged onto the map while holding the CTRL, a copy of the files is created within a path relative to the map!💗

# Scenario
I need to handle billions of files across multiple topics, for different people. And on a regular basis some of those files need to be send out to third parties, like an insurance company, mortician etc.. So I select all the necessary files that I need to get out of the map . Now I want to export the files to a path for further processing. 

# Problem
There is no file export available or any other functionality to achieve that.

# Workaround
**Fetch each file manually**
I need to repeat the following workflow for each file. 

1. Hover over node with the mouse
2. Read file path from status bar and remember path
3. Open explorer and navigate to file
4. Check if the file is exactly the file that you are looking for
5. Copy file to some temporary path

This is....   not good ;)

# Request
**Copy files to clipboard**
Put a copy of all selected files inside the clipboard. This allows the files to be pasted into some temporary path for further processing.

# Reference to discussion
https://github.com/freeplane/freeplane/discussions/1222

**Comment by macmarrum** [comment]
Here's a script that should do the trick: [copyLinkedFilesToClipboard.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/copyLinkedFilesToClipboard.groovy)

As usual, save it in `<user-dir>/scripts` and restart Freeplane.
To have it appear in `Tools->Scripts`, remove `="/menu_bar/Mac1/Copy"` from the first line.

The script works on selected nodes; it copies files from links, so that you can paste them in a file manager (e.g. Explorer).

**Comment by theelectricsoul** [comment]
@macmarrum 
Thank you for the script. But it is not working on my side. See video :) What is not seen in the video is that all the links actually do point to an existing file :)  I've checked all links. All nodes have valid link :)


https://github.com/freeplane/freeplane/assets/10260832/56f59b2d-64fb-4b77-ac29-778fea563e07

**Comment by macmarrum** [comment]
@theelectricsoul 
The script uses the mind map's parent directory to calculate the absolute path for relative links. Your mind map isn't saved, therefore the script doesn't work.

You can try again once you save your mind map.

**Comment by theelectricsoul** [comment]
@macmarrum 
There is a little mistake. The following settings are active in my Freeplane installation.

Preferences > Environment >Hyperlink types > Links = Relative

With that setting active, you cannot drop a file onto an unsaved map. Freeplane won't let you. And if you change the setting to "absolute" then you get absolutes links that work no matter if the file is saved or not. 

However ... let me demonstrate, because the user is always wrong ;)  lol :) 


https://github.com/freeplane/freeplane/assets/10260832/bc40a73c-0db8-4b38-85c4-cfa9b4bd2e75

And a second video to show that the links actually work.

https://github.com/freeplane/freeplane/assets/10260832/09852c55-27e6-4e92-84de-48bc09505040

**Comment by macmarrum** [comment]
@theelectricsoul 
But of course. I forgot about the fail-safe. You are right, a mind map must be saved if a relative link is to be added.

Thanks for the videos. I investigated some more and realized what causes the issue.
A fix is applied in [copyLinkedFilesToClipboard.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/copyLinkedFilesToClipboard.groovy). 
Please keep in mind to force-refresh (`Ctrl+F5`) the GitHub page to make sure you download the latest version.

**Comment by theelectricsoul** [comment]
@macmarrum 

See video :)  And again, the file will open in a PDF reader if clicked on within freeplane :)

https://github.com/freeplane/freeplane/assets/10260832/5eb929af-3c90-4e2f-b837-658e579784e7

**Comment by macmarrum** [comment]
@theelectricsoul 
All looks good on your side.
I ran several test cases on Windows 10 and it works OK for me.
I think you're using Window 11. I will try to get a hold of Win11 to test it there too.

In the meantime, could you please paste the copied file path(s) in Freeplane instead in Explorer to verify the generic mechanism work OK and report back?

I'm not sure if it will make any difference, but could you please try pasting the copied file path(s) in another file manager, like [Double Commander](https://doublecommander.com/) (open source) or [Total Commander](https://www.ghisler.com/) (closed source)?

**Comment by theelectricsoul** [comment]
@macmarrum 
... jesus... this is work I see. Yes itz is Win11

I am not exactly sure what you mean. But after copying the file out of the node using your script I  ..

* .. can paste into freeplane
* .. cannot paste into total Commander

https://github.com/freeplane/freeplane/assets/10260832/3202ce03-2d12-4c9e-8488-522a6d7f4bec

**Comment by macmarrum** [comment]
@theelectricsoul 

I have tested the script on Windows 11, running Freeplane 1.10.7 on Temurin Java 17. It works OK in Explorer and in Double Commander, i.e. the copied files are pasted as expected.

I'm not sure why it fails to paste for you :disappointed:

**Comment by theelectricsoul** [comment]
@macmarrum 
I was mistaken. I am running

* Microsoft Windows 10 Pro / 10.0.19045 Build 19045
* FreeplanePortable-1.11.6-pre01.paf.exe / portable Version
```batch
c:\Users\ros\Downloads>java -version
openjdk version "17.0.7" 2023-04-18 LTS
OpenJDK Runtime Environment Zulu17.42+19-CA (build 17.0.7+7-LTS)
OpenJDK 64-Bit Server VM Zulu17.42+19-CA (build 17.0.7+7-LTS, mixed mode, sharing)
```

**Comment by theelectricsoul** [comment]
@macmarrum 

![image](https://github.com/freeplane/freeplane/assets/10260832/22f79ed5-90dd-4263-b7cf-1c45b12e5544)
# After explorer
![image](https://github.com/freeplane/freeplane/assets/10260832/59dd7aff-ba39-4ffc-8640-6f4aaccce4f2)

**Comment by theelectricsoul** [comment]
@macmarrum 
Using https://www.nirsoft.net/utils/inside_clipboard.html

After using your script
```1	CF_TEXT	Memory	32	5	
7	CF_OEMTEXT	Memory	32	8	
13	CF_UNICODETEXT	Memory	64	4	
15	CF_HDROP	Memory	24	1	
16	CF_LOCALE	Memory	4	7	
49465	FileGroupDescriptor	Memory	4	3	
49466	FileGroupDescriptorW	Memory	4	2	
49475	UniformResourceLocator	Memory	32	6	
```

After using explorer
```
15	CF_HDROP	Memory	86	4	
49158	FileName	Memory	31	7	
49159	FileNameW	Memory	64	9	
49267	FileContents	Null	0	8	
49462	Shell IDList Array	Memory	477	1	
49463	Shell Object Offsets	Memory	8	11	
49466	FileGroupDescriptorW	Memory	596	10	
49471	Preferred DropEffect	Memory	4	12	
49485	DropDescription	Memory	1.044	6	
49487	AsyncFlag	Memory	4	13	
49488	UIDisplayed	Memory	4	5	
50042	DataObjectAttributes	Memory	12	2	
50043	DataObjectAttributesRequiringElevation	Memory	12	3	


```

**Comment by theelectricsoul** [comment]
@macmarrum 
Clipboardfile for the software https://freeclipboardviewer.com/
[Downloads.zip](https://github.com/freeplane/freeplane/files/12534144/Downloads.zip)

**Comment by macmarrum** [comment]
@theelectricsoul

The script works on 3 Windows machines I have access to. This makes me think that there must be something special about your machine. It's great that you keep trying.

I know that when pasting from clipboard into a file manager, the path to the file is obtained and the file manager invokes its regular copy function.

I also know that on Linux, file managers understand text/uri-list (UniformResourceLocator in InsideClipboard) and on Windows, they prefer a file list (FileGroupDescriptor in InsideClipboard) over text/uri-list.

In your clipboard viewer tools, after you run `Copy Linked File Path To Clipboard` in Freeplane, can you see the expected (absolute) path to your file in the section FileGroupDescriptor?

**Comment by theelectricsoul** [comment]
@macmarrum
FileGroupDescriptor shows the following after using your script.
![image](https://github.com/freeplane/freeplane/assets/10260832/8a29838b-f5ac-4fa7-bdfa-fe6fbeae1183)



# ChatGPT gave me this
This code works. I start it  from command line. 

![image](https://github.com/freeplane/freeplane/assets/10260832/6b269544-eaa7-485a-a043-51f273c4574c)




```
import java.awt.*;
import java.awt.datatransfer.*;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class CopyFilesToClipboard {
    public static void main(String[] args) {
        // Create a file chooser dialog
        FileDialog fileDialog = new FileDialog((Frame) null, "Select Files to Copy", FileDialog.LOAD);
        fileDialog.setMultipleMode(true); // Allow multiple file selection
        fileDialog.setVisible(true);

        // Get the selected files
        File[] selectedFiles = fileDialog.getFiles();

        if (selectedFiles.length > 0) {
            try {
                // Create a List to hold the selected files
                List<File> filesList = new ArrayList<>();
                for (File selectedFile : selectedFiles) {
                    filesList.add(selectedFile);
                }

                // Copy the selected files to the clipboard
                Toolkit toolkit = Toolkit.getDefaultToolkit();
                Clipboard clipboard = toolkit.getSystemClipboard();
                clipboard.setContents(new FileTransferable(filesList), null);

                System.out.println(selectedFiles.length + " files copied to clipboard.");
            } catch (Exception e) {
                e.printStackTrace();
            }
        } else {
            System.out.println("No files selected.");
        }
    }
}

class FileTransferable implements Transferable {
    private List<File> files;

    FileTransferable(List<File> files) {
        this.files = files;
    }

    @Override
    public DataFlavor[] getTransferDataFlavors() {
        return new DataFlavor[]{DataFlavor.javaFileListFlavor};
    }

    @Override
    public boolean isDataFlavorSupported(DataFlavor flavor) {
        return flavor.equals(DataFlavor.javaFileListFlavor);
    }

    @Override
    public Object getTransferData(DataFlavor flavor) throws UnsupportedFlavorException {
        if (flavor.equals(DataFlavor.javaFileListFlavor)) {
            return files;
        } else {
            throw new UnsupportedFlavorException(flavor);
        }
    }
}


```

**Comment by macmarrum** [comment]
@theelectricsoul 
Great investigation :+1: 

As you can probably see, the class **FileTransferable** in your Java code is virtually the same as in [copyLinkedFilesToClipboard.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/copyLinkedFilesToClipboard.groovy).

The following next investigation steps come to my mind:

1. Make sure you run you Java code with the same java.exe as you use for Freeplane. If it works OK, then the issue might be in Groovy → step 2.
2. Replace **FileTransferable** in your Java code with the one from [copyLinkedFilesToClipboard.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/copyLinkedFilesToClipboard.groovy), change the file extension to ".groovy" and run your new script with Groovy on the same JVM as you use for Freeplane.

More detailed instructions for step 2:
- Download Groovy https://groovy.jfrog.io/artifactory/dist-release-local/groovy-zips/apache-groovy-binary-4.0.13.zip
- Unzip it
- Open Terminal (PowerShell) and change to the unzipped directory
- Set JAVA_HOME to the base directory of your java.exe, which you use to run Freeplane
  `$Env:JAVA_HOME="C:\path\to\your\java\base\dir"` – this is usually one level up from `bin`, which contains `java.exe`.
- run `bin\groovy "C:\path\to\your\script.groovy"`

**Comment by theelectricsoul** [comment]
@macmarrum 
I am looking at your instructions asap :) Just a heads up, 

I tested in on two more machines. One was mine, the second one was from a friend. The only think that works is pasting the path to notepad after using your script. :)

**Comment by euu2021** [comment]
> @macmarrum I am looking at your instructions asap :) Just a heads up,
> 
> I tested in on two more machines. One was mine, the second one was from a friend. The only think that works is pasting the path to notepad after using your script. :)

This brings localization to my mind. See, for example, how the interface language affects [this script](https://docs.freeplane.org/attic/old-mediawiki-content/Batch_Jobs.html#example-exporting-to-file). And, also the slashes and back slashes [here](https://docs.freeplane.org/attic/old-mediawiki-content/Edit_scripts_in_Eclipse.html?highlight=dutch#making-available-groovy-classes-to-your-mind-map).
